packages/core/bootstrap/src/lib/middleware/cache-key/index.ts

```diff
@@ -0,0 +1,73 @@
+import type {
+  Middleware,
+  AdapterRequest,
+  Config,
+  APIEndpoint,
+  InputParameters,
+} from '@chainlink/types'
+import { Validator } from '../../modules'
+import { getHashOpts, hash, excludableInternalAdapterRequestProperties } from './util'
+import crypto from 'crypto'
+import { baseInputParameters } from '../../modules/selector'
+
+const baseInputParametersCachable = Object.keys(baseInputParameters).filter(
+  (inputParam) => !excludableInternalAdapterRequestProperties.includes(inputParam),
+)
+
+export const withCacheKey: <C extends Config>(
+  endpointSelector?: (request: AdapterRequest) => APIEndpoint<C>,
+) => Middleware = (endpointSelector) => async (execute, context) => async (input: AdapterRequest) => {
+  const endpoint = endpointSelector?.(input)
+  const cacheKey =
+    endpoint && endpoint.inputParameters
+      ? getCacheKey(input, endpoint.inputParameters)
+      : hash(input, getHashOpts()) // Fallback to legacy object hash cache key
+  const batchCacheKey =
+    endpoint && endpoint.inputParameters && endpoint.batchablePropertyPath
+      ? getBatchCacheKey(input, endpoint.inputParameters, endpoint.batchablePropertyPath)
+      : undefined
+  const inputWithCacheKey = { ...input, debug: { ...input.debug, cacheKey, batchCacheKey } }
+  return execute(inputWithCacheKey, context)
+}
+
+export function getCacheKey(request: AdapterRequest, inputParameters: InputParameters): string {
+  const inputParameterKeys = Object.keys(inputParameters ?? {}).concat(baseInputParametersCachable)
+  const validator = new Validator(request, inputParameters, {}, { shouldThrowError: false })
+
+  let data = ''
+
+  for (const key of inputParameterKeys) {
+    const value = JSON.stringify(validator.validated.data[key])
```

Comment:
just occurred to me: this wouldn't be deterministic if the validated data contains a nested object, since that order is not guaranteed